src/clients/c++/simple_shm_client.cc

```diff
@@ -0,0 +1,374 @@
+  } else if (protocol == "grpc") {
+    err = nic::ServerHealthGrpcContext::Create(&health_ctx, url, verbose);
+  } else {
+    Usage(argv, "unknown protocol '" + protocol + "'");
+  }
+
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to create health context: " << err << std::endl;
+    exit(1);
+  }
+
+  bool live, ready;
+  err = health_ctx->GetLive(&live);
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to get server liveness: " << err << std::endl;
+    exit(1);
+  }
+
+  err = health_ctx->GetReady(&ready);
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to get server readiness: " << err << std::endl;
+    exit(1);
+  }
+
+  std::cout << "Health for model " << model_name << ":" << std::endl;
+  std::cout << "Live: " << live << std::endl;
+  std::cout << "Ready: " << ready << std::endl;
+
+  // Create a status context and get the status of the model.
+  std::unique_ptr<nic::ServerStatusContext> status_ctx;
+  if (protocol == "http") {
+    err = nic::ServerStatusHttpContext::Create(
+        &status_ctx, url, http_headers, model_name, verbose);
+  } else if (protocol == "grpc") {
+    err = nic::ServerStatusGrpcContext::Create(
+        &status_ctx, url, model_name, verbose);
+  } else {
+    Usage(argv, "unknown protocol '" + protocol + "'");
+  }
+
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to create status context: " << err << std::endl;
+    exit(1);
+  }
+
+  ni::ServerStatus server_status;
+  err = status_ctx->GetServerStatus(&server_status);
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to get status: " << err << std::endl;
+    exit(1);
+  }
+
+  std::cout << "Status for model " << model_name << ":" << std::endl;
+  std::cout << server_status.DebugString() << std::endl;
+
+  // Create the inference context for the model.
+  std::unique_ptr<nic::InferContext> infer_ctx;
+  if (protocol == "http") {
+    err = nic::InferHttpContext::Create(
+        &infer_ctx, url, http_headers, model_name, -1 /* model_version */,
+        verbose);
+  } else if (protocol == "grpc") {
+    err = nic::InferGrpcContext::Create(
+        &infer_ctx, url, model_name, -1 /* model_version */, verbose);
+  } else {
+    Usage(argv, "unknown protocol '" + protocol + "'");
+  }
+
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to create inference context: " << err
+              << std::endl;
+    exit(1);
+  }
+
+  // Set the context options to do batch-size 1 requests. Also request
+  // that all output tensors be returned.
+  std::unique_ptr<nic::InferContext::Options> options;
+  FAIL_IF_ERR(
+      nic::InferContext::Options::Create(&options),
+      "unable to create inference options");
+
+  options->SetBatchSize(1);
+  for (const auto& output : infer_ctx->Outputs()) {
+    options->AddRawResult(output);
+  }
+
+  FAIL_IF_ERR(
+      infer_ctx->SetRunOptions(*options), "unable to set inference options");
+
+  // Create the data for the two input tensors. Initialize the first
```

Comment:
Can you allocate and create the data directly in the shared memory. Also it will be more informative if you use a single shared memory region (or perhaps one for inputs and one for outputs).

Reply:
Yes, I'll allocate the data directly in shared memory.
I kept two different shared memory regions (1 for each input tensor name). Similarly for the output tensors.